[21.05] discourse: 2.7.5 -> 2.7.7, plugins and fixes

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

- Make recent plugin additions and a change to the builder available in stable.
- Update `discourse` to the latest stable version
- Update the plugins
- Run tests on the correct package, i.e. `discourseAllPlugins.tests` now runs the tests on the `discourseAllPlugins` package
- Change the auto generated plugin assets path to one that is writable at asset build time (fixes the issue in #132475)
- Make the `update.py` script able to package plugins from scratch

Closes #132113

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Stunkymonkey committed:
discourse.mkDiscoursePlugin: remove phases

(cherry picked from commit 4f62b051375b0fdaa01ba37e915cfaaa7de40d80)

ryantm committed:
discourseAllPlugins: add discourse-calendar

(cherry picked from commit 858b0157e56426ee0d56faae393fc85b361ffd90)

ryantm committed:
discourseAllPlugins: init discourse-checklist

(cherry picked from commit eb02dc0c4a14d5a15f3443567bc12abf631c300a)

ryantm committed:
discourseAllPlugins: init discourse-data-explorer

(cherry picked from commit 601db31c264f3008c25f3fb13f1968a543a9cfe6)

ryantm committed:
discourseAllPlugins: init discourse-migratepassword

(cherry picked from commit 85d7eb75c968421ea233f4c8a5b05886733c2dec)

ryantm committed:
discourse/update.py: add missing plugins

(cherry picked from commit 32e24e792caefa1de46bf20da2ba4d8fbbf035dc)

talyz committed:
discourse: 2.7.5 -> 2.7.7

(cherry picked from commit c97ae4ad8e782765b333840d8d1aa83903f82b7f)

talyz committed:
discourse.plugins.discourse-calendar: Update

(cherry picked from commit d1a63bf19c6c75430393dab94f5e939b258ffdb6)

talyz committed:
discourse.plugins.discourse-canned-replies: Update

(cherry picked from commit 5d94e3bfc94637f6876b9957471fc571646867b0)

talyz committed:
discourse.plugins.discourse-solved: Update

(cherry picked from commit 92b758266e61b2efe078973a3d020b3d1a2c363d)

talyz committed:
discourse.plugins.discourse-data-explorer: Update

(cherry picked from commit 6f2652735817e22c55de3b6e137faf7cdbc3fd2a)

talyz committed:
discourse: Remove leftover link to unused plugins directory

(cherry picked from commit bb14315d51a1c35304dfd82ff5c1faeaaf2c450c)

talyz committed:
discourse.tests: Test the appropriate discourse package

Perform the tests on the package that the `tests` attribute is a child
of, i.e. if `discourseAllPlugins.tests` is built, the tests will run
with the `discourseAllPlugins` package, not the `discourse` package as
previously.

(cherry picked from commit 6fd5a40ccaf0b4da1362803a387bf46d381dd66a)

talyz committed:
discourse: Change the path to the auto generated plugin assets

Change the path to the auto generated plugin assets, which defaults to
the plugin's directory and isn't writable at the time of asset
generation.

(cherry picked from commit 443b318ee9c614d480a2ecb0120b52806d6fbb3b)

r-rmcgibbo commented:
Result of `nixpkgs-review pr 133624` at c1305b48 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>discourseAllPlugins</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `meta`.
Near pkgs/servers/web-apps/discourse/default.nix:56:7:
```
   |
56 |     , meta ? null
   |       ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

bundler is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/servers/web-apps/discourse/default.nix:218:5:
```
    |
218 |     buildInputs = [
    |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `replace`.
Near pkgs/servers/web-apps/discourse/default.nix:4:9:
```
  |
4 | , ruby, replace, gzip, gnutar, git, cacert, util-linux, gawk
  |         ^
```
</li>
    <li>warning: unused-argument

Unused argument: `src`.
Near pkgs/servers/web-apps/discourse/default.nix:58:7:
```
   |
58 |     , src
   |       ^
```
</li>
  </ul>
</details>

talyz committed:
discourse.plugins.discourse-github: Update

(cherry picked from commit 4197b6dd146c395f1caec3b086334a2b0eff623a)

talyz committed:
discourse.plugins: Make the updater able to package plugins

Let the update.py script handle the initial, repetitive task of
packaging new plugins. With this in place, the plugin only needs to be
added to the list in `update-plugins` and most of the work will be
done automatically when the script is run. Metadata still needs to be
filled in manually and some packages may of course require additional
work/patching.

(cherry picked from commit f8096460bd15d4f13a01cfddf0a30798921fdb42)

talyz committed:
discourse.mkDiscoursePlugin: Handle repos with `gems` directories

Some plugin repos already have a `gems` directory. This lets the
packager choose whether it should be kept and the nix packaged ruby
gems should be copied into it or if it should be removed in favor of
our ruby gems.

(cherry picked from commit 04e6b03fa91603c7f1961cfcdcf5880c91fe6b05)

talyz committed:
discourse: update.py: Remove native platforms in plugin lock files..

...and add ruby.

(cherry picked from commit 12ff4b79e48d9c4aa5660d2ec7ce52a21806d8e4)
